watchers/manager.go

```diff
@@ -0,0 +1,148 @@
+	RemoveExpiredWatchers() error
+
+	// GetWatcher returns the Watcher instance with the specified name
+	GetWatcher(name string) Watcher
+}
+
+// Watchers represents a map of Watcher instances
+type Watchers map[string]Watcher
+
+// manager is a struct that implement the Manager interface.
+type manager struct {
+	// controller is used to take event of a source send them to a watcher
+	controller controller.Controller
+
+	// reader used to communicate with the k8s api. It allows us to check the existence of resources in k8s
+	reader client.Reader
+
+	// watchers represents a map of Watcher instances
+	watchers Watchers
+}
+
+// GetWatcher returns the Watcher instance with the given name
+func (m manager) GetWatcher(name string) Watcher {
+	return m.watchers[name]
+}
+
+// NewManager creates a new Manager instance
+func NewManager(r client.Reader, c controller.Controller) Manager {
+	return manager{
+		controller: c,
+		reader:     r,
+		watchers:   Watchers{},
+	}
+}
+
+// AddWatcher adds a new Watcher instance to the Manager
+func (m manager) AddWatcher(w Watcher) error {
+	watcherName := w.GetName()
+
+	// Check if the Watcher instance already exists in the Manager
+	if ok := m.watchers[watcherName]; ok != nil {
+		return nil
+	}
+
+	// Start the Watcher instance
+	if err := w.Start(); err != nil {
+		return fmt.Errorf("could not start the watcher. Error: %s", err)
+	}
+
+	// Add the Watcher instance to the Manager's Watchers
+	m.watchers[watcherName] = w
+
+	// Get the cache source of the Watcher instance
+	cacheSource, err := w.GetCacheSource()
+	if err != nil {
+		return fmt.Errorf("could not get the cache source of the watcher. Error: %s", err)
+	}
+
+	// Get the context tuple of the Watcher instance
+	ctxTuple, _ := w.GetContextTuple()
+
+	// Watch for changes in the cache
+	return m.controller.Watch(
+		cacheSource,
+		handler.EnqueueRequestsFromMapFunc(
+			func(c client.Object) []reconcile.Request {
+				return []reconcile.Request{{NamespacedName: ctxTuple.DisruptionNamespacedName}}
+			},
+		),
+	)
+}
+
+// RemoveWatcher removes a Watcher instance from the Manager
+func (m manager) RemoveWatcher(w Watcher) error {
+	watcherName := w.GetName()
+
+	// Check if the Watcher instance exists in the Manager
+	if ok := m.watchers[watcherName]; ok == nil {
+		return fmt.Errorf("the watcher %s does not exist", watcherName)
+	}
+
+	// Stop and delete the Watcher instance
+	w.Clean()
+	delete(m.watchers, watcherName)
+
+	return nil
+}
+
+// RemoveExpiredWatchers removes any expired Watcher instances from the Manager
+func (m manager) RemoveExpiredWatchers() error {
+	for name, w := range m.watchers {
+		if w.IsExpired() {
+			w.Clean()
+			delete(m.watchers, name)
+
+			continue
+		}
+
+		m.deleteOrphanWatcher(w)
+	}
+
+	return nil
+}
+
+// deleteOrphanWatcher deletes a Watcher instance if the associated Disruption object does not exist
+func (m manager) deleteOrphanWatcher(watcher Watcher) bool {
```

Comment:
This feels to me it may not be the same thing as expired watchers

Why hiding this method instead of providing it in the interface and being able to schedule both separately?

Reply:
Good point! I will change that to 👍